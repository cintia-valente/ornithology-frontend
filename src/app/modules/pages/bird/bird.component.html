<app-nav-bar></app-nav-bar>
<div class="birds">
  <div class="container">
    <div class="row">
      <div class="col-md-1">
        <div class="title">Aves</div>
      </div>

      <div class="col-md-1">
        <div class="image">
          <img src="assets/img/logo3.png" style="width: 70px" />
        </div>
      </div>
    </div>

    <div class="search-container">
      <form>
        <div class="form-group">
          <input
            type="text"
            class="form-control"
            placeholder="Buscar"
            (input)="search($event)"
          />
        </div>
      </form>
    </div>
    <div class="row">
      <div class="col-md-3" *ngFor="let bird of birds">
        <a
          class="nav-item nav-link"
          [routerLink]="['../cadastrar/', bird.namePtbr]"
        >
          <div class="card">
            <div class="card-body">
              <img [src]="bird.picByte" *ngIf="bird.imageId" />
              <div class="bird-namePtbr">
                {{ bird.namePtbr }}
              </div>
              <div class="bird-nameEnglish">
                {{ bird.nameEnglish }}
              </div>
              <div class="bird-nameLatin">
                {{ bird.namePtbr }}
              </div>
              <div class="bird-size">
                {{ bird.size }}
              </div>
              <div class="bird-genre">
                {{ bird.genre }}
              </div>
              <div class="bird-color">
                {{ bird.color }}
              </div>
              <div class="bird-family">
                {{ bird.family }}
              </div>
              <div class="bird-habitat">
                {{ bird.habitat }}
              </div>
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
