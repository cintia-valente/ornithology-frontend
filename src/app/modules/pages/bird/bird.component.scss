@import "/src/app/scss/variables.scss";

.title {
    font-size: 30px;
    margin-top: 9px;
    font-style: italic;
    font-weight: 500px;
    color: $black;
    margin-bottom: inherit;
}

.card {
    transition: all 0.3s;
    color: initial;
    height: 100%;
    border-radius: 10px;

    img {
        border-radius: 10px 10px 0px 0px;;
    }
    
    &:hover {
        cursor: pointer;
        box-shadow: 3px 3px 3px 3px $primary-border-color;
    }
}