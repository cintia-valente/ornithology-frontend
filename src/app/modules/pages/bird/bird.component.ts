import { FileService } from './../../../services/file.service';
import { HttpErrorResponse } from '@angular/common/http';
import { Component, OnInit } from '@angular/core';
import { DomSanitizer, SafeResourceUrl } from '@angular/platform-browser';
import { ToastrService } from 'ngx-toastr';
import { throwError } from 'rxjs';
import { Bird } from 'src/app/model/bird.model';
import { BirdService } from '../../../services/bird.service';
import { Buffer } from 'buffer';

@Component({
  selector: 'app-bird',
  templateUrl: './bird.component.html',
  styleUrls: ['./bird.component.scss'],
})
export class BirdComponent implements OnInit {
  birds: Bird[] = [];
  birdsDisplayed: Bird[] = [];
  error: boolean = false;
  imageUrl: any;

  constructor(
    private birdService: BirdService,
    private fileService: FileService,
    private toastr: ToastrService,
    private sanitizer: DomSanitizer
  ) {}

  ngOnInit(): void {
    this.listBirds();
  }

  listBirds() {
    this.error = false;

    this.birdService.getBirds().subscribe({
      next: (data: Bird[]) => {
        this.birds = data;
        this.birdsDisplayed = data;

        this.listFiles();
      },

      error: (err: HttpErrorResponse) => {
        this.toastr.error(
          'Erro ao carregar aves. Por favor, tente novamente mais tarde.'
        );

        return throwError(() => err);
      },
    });
  }

  listFiles() {
    this.birds.forEach((bird) => {
      if (bird.imageId) {
        this.fileService
          .getFiles(bird.imageId, { responseType: 'blob' })
          .subscribe((data) => {
            if (data) {
              bird.picByte = URL.createObjectURL(data);
            }
          });
      }
    });
  }

  search(event: Event): void {
    const target = event.target as HTMLInputElement;
    const value = target.value;

    this.birds = this.birdsDisplayed.filter((bird) =>
      bird.namePtbr.toLowerCase().includes(value)
    );
  }
}
