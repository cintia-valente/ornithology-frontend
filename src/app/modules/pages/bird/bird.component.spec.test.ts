import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import {
  ComponentFixture,
  getTestBed,
  TestBed,
  TestComponentRenderer,
} from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { Routes } from '@angular/router';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';

import { ToastrService } from 'ngx-toastr';
import { of } from 'rxjs';

import { BirdService } from '../../../services/bird.service';

import { BirdComponent } from './bird.component';

const mockBirds = [
  {
    id: '27',
    image: 'img',
    namePtbr: 'Arara',
    nameEnglish: 'Blue-and-yellow Macaw',
    nameLatin: 'Ara ararauna',
    size: '90 cm',
    genre: 'Can',
    color: 'Amarelo',
    family: 'Psittacidae',
    habitat: 'Floresta',
  },
];

class BirdServiceMock {
  getBirds() {
    return of();
  }
}

class ToastrServiceMock {
  success(message?: string) {}

  error(message?: string) {}
}

const routes: Routes = [
  {
    path: 'cadastrar/:birdName',
    component: TestComponentRenderer,
  },
];

describe('BirdComponent', () => {
  let injector: TestBed;
  let component: BirdComponent;
  let fixture: ComponentFixture<BirdComponent>;
  let birdService: BirdService;
  let toastrService: ToastrService;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        CommonModule,
        HttpClientModule,
        RouterTestingModule,
        RouterTestingModule.withRoutes(routes),
      ],
      declarations: [BirdComponent],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      providers: [
        { provide: BirdService, useClass: BirdServiceMock },
        { provide: ToastrService, useClass: ToastrServiceMock },
      ],
    }).compileComponents();
  });

  beforeEach(() => {
    injector = getTestBed();
    fixture = TestBed.createComponent(BirdComponent);

    birdService = injector.inject(BirdService);
    toastrService = injector.inject(ToastrService);

    jest.spyOn(birdService, 'getBirds').mockReturnValue(of(mockBirds));

    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it(`Dado: que o componente foi carregado
      Então: deve chamar o serviço birdService.getBirds`, async () => {
    //Arrange
    const spyUserList = jest
      .spyOn(birdService, 'getBirds')
      .mockReturnValue(of(mockBirds));

    //Act
    fixture.detectChanges();

    //Assert
    expect(spyUserList).toHaveBeenCalled();
  });

  it(`Dado: que o componente foi carregado
      Quando: detectar as alterações
      Então: deve gerar um Snapshot do componente carregado`, async () => {
    //Arrange
    jest.spyOn(birdService, 'getBirds').mockReturnValue(of(mockBirds));

    //Act
    fixture.detectChanges();

    //Assert
    expect(fixture).toMatchSnapshot();
  });
});
