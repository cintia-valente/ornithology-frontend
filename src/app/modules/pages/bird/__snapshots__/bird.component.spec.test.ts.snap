// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BirdComponent Dado: que o componente foi carregado
      Quando: detectar as alterações
      Então: deve gerar um Snapshot do componente carregado 1`] = `
<app-bird
  __ngContext__={[Function Number]}
  birdService={[Function BirdServiceMock]}
  birds={[Function Array]}
  birdsDisplayed={[Function Array]}
  error="false"
  fileService={[Function FileService]}
  sanitizer={[Function DomSanitizerImpl]}
  toastr={[Function ToastrServiceMock]}
>
  <app-nav-bar /><div
    class="birds"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-1"
        >
          <div
            class="title"
          >
            Aves
          </div>
        </div>
        <div
          class="col-md-1"
        >
          <div
            class="image"
          >
            <img
              src="assets/img/logo3.png"
              style="width: 70px;"
            />
          </div>
        </div>
      </div>
      <div
        class="search-container"
      >
        <form>
          <div
            class="form-group"
          >
            <input
              class="form-control"
              placeholder="Buscar"
              type="text"
            />
          </div>
        </form>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          <a
            class="nav-item nav-link"
            href="/cadastrar/Arara"
          >
            <div
              class="card"
            >
              <div
                class="card-body"
              >
                <div
                  class="bird-namePtbr"
                >
                   Arara 
                </div>
                <div
                  class="bird-nameEnglish"
                >
                   Blue-and-yellow Macaw 
                </div>
                <div
                  class="bird-nameLatin"
                >
                   Arara 
                </div>
                <div
                  class="bird-size"
                >
                   90 cm 
                </div>
                <div
                  class="bird-genre"
                >
                   Can 
                </div>
                <div
                  class="bird-color"
                >
                   Amarelo 
                </div>
                <div
                  class="bird-family"
                >
                   Psittacidae 
                </div>
                <div
                  class="bird-habitat"
                >
                   Floresta 
                </div>
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</app-bird>
`;
